### PR TITLE
[jwt] Fetch RSA public key from key_server

### PR DESCRIPTION
**Pull request by Harshg999**
## What changes were proposed in this pull request?

- Fetch RSA public key from the key_server to be used in decoding of JWT.
- If verify flag is False, then it will not be fetched.
- Update unit tests for patching request call.

### CODE REVIEW

**Comment by Harshg999** [comment]
Still need to clarify what is the actual response format containing key and in which format is the key is being passed to us @romainr

File: desktop/core/src/desktop/auth/api_authentications.py [hunk 1]
```diff
@@ -49,10 +50,15 @@ def authenticate(self, request):
 
     LOG.debug('JwtAuthentication: got access token %s' % access_token)
 
+    # Fetch public key
+    token_metadata = jwt.get_unverified_header(access_token)
+    headers = {'kid': token_metadata.get('kid')} if token_metadata.get('kid') is not None else {}
```
Comment (romainr) [inline desktop/core/src/desktop/auth/api_authentications.py:55]: Often providing a default value for the get is simpler:

`token_metadata.get('kid', {})`

File: desktop/core/src/desktop/auth/api_authentications.py [hunk 2]
```diff
@@ -49,10 +50,15 @@ def authenticate(self, request):
 
     LOG.debug('JwtAuthentication: got access token %s' % access_token)
 
+    # Fetch public key
+    token_metadata = jwt.get_unverified_header(access_token)
```
Comment (romainr) [inline desktop/core/src/desktop/auth/api_authentications.py:54]: I think you could just use the json they provided to and mock it back in a dedicated unit test

**Review by romainr — APPROVED** [review]

File: desktop/core/src/desktop/auth/api_authentications_tests.py [hunk 3]
```diff
@@ -48,73 +48,113 @@ def setUp(self):
...
+
+        reset = AUTH.JWT.VERIFY.set_for_testing(True)
+        try:
+          user, token = JwtAuthentication().authenticate(request=self.request)
+
+          jwt_decode.assert_called_with(
+            self.sample_token,
+            b'-----BEGIN PUBLIC KEY-----\nMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEArtT3gR0NDIx6gv8xYLiP\nue/ItaIbognCGGgQbipp3IOuobu2RnJjedsIRBTEOdkVx+xjV6m92VYtrpW6gM9v\nldwTfI0UmoSLGKT5uYd0JGHvYWoN9inCZYZcnala58T8HDgLiXa9KlEuQxGGQDem\nB3yf5rgS1OhLBKVsI8bMVgah7xNIiBOWsVeWIEr13Nem8HUuDqgIpL/8TgjxFOqF\ncdqPCfoIZ89JKEiKbsGbU+lqs1xYChFscI/w7Jc7l6rvf2nsLGMFs3U4ZJvS4AUp\nVno2e527clXzQisfJKwb4hjfKRMhHfnYfyJxaoHqWfx8DjXmH3CMqlWr/+hL3y1+\n4QIDAQAB\n-----END PUBLIC KEY-----\n',
+            algorithms=['RS256'],
+            verify=True
```
Comment (romainr) [inline desktop/core/src/desktop/auth/api_authentications_tests.py:156]: :+1:

File: desktop/core/src/desktop/auth/api_authentications.py [hunk 4]
```diff
@@ -49,10 +52,17 @@ def authenticate(self, request):
 
     LOG.debug('JwtAuthentication: got access token from %s: %s' % (request.path, access_token))
 
+    # Handle public key
+    public_key_pem = ''
+    if AUTH.JWT.VERIFY.get():
+      jwk = requests.get(AUTH.JWT.KEY_SERVER_URL.get())
+      public_key = jwt.algorithms.RSAAlgorithm.from_jwk(json.dumps(jwk["keys"][0])).public_key()
+      public_key_pem = public_key.public_bytes(encoding=serialization.Encoding.PEM,format=serialization.PublicFormat.SubjectPublicKeyInfo)
```
Comment (romainr) [inline desktop/core/src/desktop/auth/api_authentications.py:60]: nit: space post comma
`Encoding.PEM, `

File: desktop/core/src/desktop/auth/api_authentications.py [hunk 5]
```diff
@@ -49,10 +52,17 @@ def authenticate(self, request):
 
     LOG.debug('JwtAuthentication: got access token from %s: %s' % (request.path, access_token))
 
+    # Handle public key
+    public_key_pem = ''
+    if AUTH.JWT.VERIFY.get():
+      jwk = requests.get(AUTH.JWT.KEY_SERVER_URL.get())
+      public_key = jwt.algorithms.RSAAlgorithm.from_jwk(json.dumps(jwk["keys"][0])).public_key()
```
Comment (romainr) [inline desktop/core/src/desktop/auth/api_authentications.py:59]: This will 500 blow-up if we need proper auth or json is improper, but good enough for v1

Comment (romainr) [reply desktop/core/src/desktop/auth/api_authentications.py:59]: Maybe check `AUTH.JWT.KEY_SERVER_URL.get()` is not empty, maybe same if jwk.get('keys') is empty

**Review by romainr — APPROVED** [review]

File: desktop/core/src/desktop/auth/api_authentications.py [hunk 6]
```diff
@@ -82,6 +89,19 @@ def authenticate(self, request):
 
     return (user, None)
 
+  def _handle_public_key(self, access_token):
+    token_metadata = jwt.get_unverified_header(access_token)
+    headers = {'kid': token_metadata.get('kid', {})}
```
Comment (Harshg999) [inline desktop/core/src/desktop/auth/api_authentications.py:94]: @wing2fly

**Pull request merged**